File: TriCoachAI/Models/Workout.swift
import Foundation

/// A target intensity, expressed however is most natural for the sport
/// (HR zone always available; pace/power/cadence when relevant).
struct TargetZone: Codable, Equatable, Hashable {
    var hrZone: Int?
    var hrRangeBpm: ClosedRangeCodable<Int>?
    var paceSecPerKm: ClosedRangeCodable<Int>?
    var paceSecPer100m: ClosedRangeCodable<Int>?
    var powerWatts: ClosedRangeCodable<Int>?
    var cadence: Int?
}

/// A single block within the main set (e.g. "5 x 3 min @ Z4, récup 2 min Z1").
struct IntervalBlock: Codable, Equatable, Hashable, Identifiable {
    var id: UUID = UUID()
    var repetitions: Int = 1
    var workDurationSec: Int
    var recoveryDurationSec: Int = 0
    var target: TargetZone
    var note: String?
}

struct WorkoutSection: Codable, Equatable, Hashable {
    var durationMin: Int
    var description: String
    var target: TargetZone
}

struct WorkoutStructure: Codable, Equatable, Hashable {
    var warmup: WorkoutSection
    var mainSet: [IntervalBlock]
    var cooldown: WorkoutSection
}

struct Workout: Codable, Equatable, Hashable, Identifiable {
    var id: UUID = UUID()
    var date: Date
    var sport: SportType
    var title: String
    var summary: String
    var structure: WorkoutStructure
    var plannedDurationMin: Int
    var plannedDistanceM: Int?
    var estimatedTSS: Double?
    var estimatedTRIMP: Double?
    var rpeTarget: Int?
    var intensity: WorkoutIntensity
    var status: WorkoutStatus = .planned
    var calendarEventId: String?
    var isRecoveryWeek: Bool = false
}

/// `ClosedRange` isn't directly `Codable`; this thin wrapper is.
struct ClosedRangeCodable<Bound: Codable & Comparable & Hashable>: Codable, Equatable, Hashable {
    var lowerBound: Bound
    var upperBound: Bound

    init(_ range: ClosedRange<Bound>) {
        lowerBound = range.lowerBound
        upperBound = range.upperBound
    }

    init(lowerBound: Bound, upperBound: Bound) {
        self.lowerBound = lowerBound
        self.upperBound = upperBound
    }

    var range: ClosedRange<Bound> { lowerBound...upperBound }
}

File: TriCoachAndroid/app/src/main/java/com/tricoach/android/models/Workout.kt
package com.tricoach.android.models

import kotlinx.serialization.Serializable

@Serializable
data class ValueRange(
    val lowerBound: Double,
    val upperBound: Double,
)

@Serializable
data class TargetZone(
    val hrZone: Int? = null,
    val hrRangeBpm: ValueRange? = null,
    val paceSecPerKm: ValueRange? = null,
    val paceSecPer100m: ValueRange? = null,
    val powerWatts: ValueRange? = null,
    val cadence: Int? = null,
)

@Serializable
data class WorkoutSection(
    val durationMin: Int,
    val description: String,
    val target: TargetZone,
)

@Serializable
data class IntervalBlock(
    val id: String,
    val repetitions: Int,
    val workDurationSec: Int,
    val recoveryDurationSec: Int,
    val target: TargetZone,
    val note: String? = null,
)

@Serializable
data class WorkoutStructure(
    val warmup: WorkoutSection,
    val mainSet: List<IntervalBlock>,
    val cooldown: WorkoutSection,
)

@Serializable
data class Workout(
    val id: String,
    val date: String,
    val sport: SportType,
    val title: String,
    val summary: String,
    val structure: WorkoutStructure,
    val plannedDurationMin: Int,
    val plannedDistanceM: Double? = null,
    val estimatedTSS: Double? = null,
    val estimatedTRIMP: Double? = null,
    val rpeTarget: Int? = null,
    val intensity: WorkoutIntensity,
    val status: WorkoutStatus,
    val calendarEventId: String? = null,
    val isRecoveryWeek: Boolean = false,
)
